File: simulations-java/simulations/circuit-construction-kit/src/edu/colorado/phet/circuitconstructionkit/model/mna/TestPrototype.java
// Copyright 2002-2011, University of Colorado
package edu.colorado.phet.circuitconstructionkit.model.mna;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.DecimalFormat;

public class TestPrototype {
    public static double euclideanDistance(double x, double y) {
        return Math.sqrt((x - y) * (x - y));
    }

    public static CapacitorState updateWithSubdivisions(final double voltage, final double resistance, final double capacitance, CapacitorState capacitorState, double dt) {
        TimestepSubdivisions.Steppable<CapacitorState> steppable = new TimestepSubdivisions.Steppable<CapacitorState>() {
            public double distance(CapacitorState a, CapacitorState b) {
                return euclideanDistance(a.current, b.current); //a.distance(b)//TODO: improve distance metric; just using current euclidean for comparison for TestCompanionModel
            }

            public CapacitorState update(CapacitorState a, double dt) {
                return TestPrototype.updateIt(voltage, resistance, capacitance, a, dt);
            }
        };
        return new TimestepSubdivisions<CapacitorState>(1E-7, 1E-8).stepInTime(capacitorState, steppable, dt);
    }

    public static CapacitorState updateIt(double voltage, double resistance, double capacitance, CapacitorState state, double dt) {
        //TRAPEZOIDAL
        double companionResistance = dt / 2 / capacitance;
        double companionBatteryVoltage = state.voltage + companionResistance * state.current;

        //    println("companion resistor resistance = "+companionResistorResistance+", companion battery voltage = "+companionBatteryVoltage)

        //BACKWARD EULER
        //    val companionBatteryVoltage = state.voltage
        //    val companionResistorResistance = dt / capacitance

        double newCurrent = (voltage - companionBatteryVoltage) / (companionResistance + resistance);
        double newVoltage = voltage - newCurrent * resistance; //signs may be wrong here

        return new CapacitorState(newVoltage, newCurrent);
    }

    //voltage and current across the capacitor
    static class CapacitorState {
        double voltage;
        double current;

        public CapacitorState(double voltage, double current) {
            this.voltage = voltage;
            this.current = current;
        }
    }

    public static void main(String[] args) throws IOException {
        DecimalFormat f = new DecimalFormat("0.000000000000000");
        double voltage = 9;
//        double resistance = 1;
        double resistance = 1E-6;
        double capacitance = 0.1;
        double time = 0.0;
        double dt = 0.03;
        CapacitorState state = new CapacitorState(0, voltage / resistance);

        String headers = "iteration \t dt \t t \t v(t) \t i(t) \t vTrue \t vNumerical \t error";
        System.out.println(headers);

        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(new File("C:/Users/Owner/Desktop/cck-out" + System.currentTimeMillis() + ".txt")));
        bufferedWriter.write(headers + "\n");

        for (int j = 0; j < 15; j++) {
            double vTrue = voltage * Math.exp(-time / resistance / capacitance);
            double vNumeric = voltage - state.voltage;

            double error = Math.abs(vTrue - vNumeric);
            String str = j + "\t" + f.format(dt) + "\t" + f.format(time) + "\t" + f.format(state.voltage) + "\t" + f.format(state.current) + "\t" + f.format(vTrue) + "\t" + f.format(vNumeric) + "\t" + f.format(error);
            System.out.println(str);
            bufferedWriter.write(str + "\n");

            state = updateWithSubdivisions(voltage, resistance, capacitance, state, dt);
            time = time + dt;
        }
        bufferedWriter.close();
    }
}

File: branches/simsharing-mongo/simulations-java/simulations/circuit-construction-kit/scala-src/edu/colorado/phet/cckscala/tests/TestPrototype.scala
package edu.colorado.phet.cckscala.tests

import java.text.DecimalFormat
import java.io.{File, FileWriter, BufferedWriter}

object TestPrototype {
  def updateWithSubdivisions(voltage: Double, resistance: Double, capacitance: Double, capacitorState: CapacitorState, dt: Double) = {
    val steppable = new Steppable[CapacitorState] {
      def update(a: CapacitorState, dt: Double) = TestPrototype.update(voltage, resistance, capacitance, a, dt)

      def distance(a: CapacitorState, b: CapacitorState) = MathUtil.euclideanDistance(a.current :: Nil, b.current :: Nil) //a.distance(b)//TODO: improve distance metric; just using current euclidean for comparison for TestCompanionModel
    }
    new TimestepSubdivisions(1E-7).stepInTime(capacitorState, steppable, dt)
  }

  def update(voltage: Double, resistance: Double, capacitance: Double, state: CapacitorState, dt: Double) = {
    //TRAPEZOIDAL
    val companionBatteryVoltage = state.voltage + dt / 2 / capacitance * state.current
    val companionResistorResistance = dt / 2 / capacitance

    //    println("companion resistor resistance = "+companionResistorResistance+", companion battery voltage = "+companionBatteryVoltage)

    //BACKWARD EULER
    //    val companionBatteryVoltage = state.voltage
    //    val companionResistorResistance = dt / capacitance

    val newCurrent = (voltage - companionBatteryVoltage) / (companionResistorResistance + resistance)
    val newVoltage = voltage - newCurrent * resistance //signs may be wrong here

    new CapacitorState(newVoltage, newCurrent)
  }

  //voltage and current across the capacitor
  case class CapacitorState(voltage: Double, current: Double)

  def main(args: Array[String]) {
    val f = new DecimalFormat("0.000000000000000")
    val voltage = 9
    //    val resistance = 1
    val resistance = 1E-6
    val capacitance = 0.1
    var time = 0.0
    val dt = 0.03
    var state = new CapacitorState(0, voltage / resistance)

    val headers = "iteration \t dt \t t \t v(t) \t i(t) \t vTrue \t vNumerical \t error"
    println(headers)

    val bufferedWriter = new BufferedWriter(new FileWriter(new File("C:/Users/Owner/Desktop/cck-out" + System.currentTimeMillis() + ".txt")))
    bufferedWriter.write(headers + "\n")

    for (j <- 0 until 15) {
      val vTrue = voltage * Math.exp(-time / resistance / capacitance)
      val vNumeric = voltage - state.voltage

      val error = Math.abs(vTrue - vNumeric)
      val str = j + "\t" + f.format(dt) + "\t" + f.format(time) + "\t" + f.format(state.voltage) + "\t" + f.format(state.current) + "\t" + f.format(vTrue) + "\t" + f.format(vNumeric) + "\t" + f.format(error)
      println(str)
      bufferedWriter.write(str + "\n")

      state = updateWithSubdivisions(voltage, resistance, capacitance, state, dt)
      time = time + dt
    }
    bufferedWriter.close()
  }
}